File: typescript/src/integrations/crewai.ts
/**
 * CrewAI integration for AgentSentinel.
 *
 * Wraps CrewAI tools and agent actions with AgentSentinel protection.
 *
 * @example
 * ```ts
 * import { AgentPolicy, AgentGuard } from "@agentsentinel/sdk";
 * import { CrewAIGuard, protectCrew } from "@agentsentinel/sdk/integrations/crewai";
 *
 * const policy = new AgentPolicy({ dailyBudget: 10.0, requireApproval: ["send_email"] });
 * const guard = new AgentGuard({ policy });
 * const crewaiGuard = new CrewAIGuard(guard);
 *
 * const searchWeb = crewaiGuard.tool(
 *   (query: string) => `Results for ${query}`,
 *   { name: "search_web", cost: 0.01 }
 * );
 * ```
 */

import { AgentGuard, ProtectOptions } from "../guard";
import { AgentPolicy } from "../policy";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFn = (...args: any[]) => any;

export interface CrewAIToolOptions {
  /** Override tool name (defaults to function name). */
  name?: string;
  /** Fixed cost per call (USD). */
  cost?: number;
  /** Model name for cost tracking. */
  model?: string;
}

/**
 * Wraps CrewAI tools and agents with {@link AgentGuard} policy enforcement.
 */
export class CrewAIGuard {
  private readonly _guard: AgentGuard;

  constructor(guardOrPolicy: AgentGuard | AgentPolicy) {
    this._guard =
      guardOrPolicy instanceof AgentGuard
        ? guardOrPolicy
        : new AgentGuard({ policy: guardOrPolicy });
  }

  /** Access the underlying {@link AgentGuard}. */
  get guard(): AgentGuard {
    return this._guard;
  }

  /**
   * Wrap a tool function with policy enforcement.
   *
   * @param fn   The tool function to protect.
   * @param opts Optional overrides for name, cost, and model.
   * @returns    Protected function with the same signature.
   */
  tool<T extends AnyFn>(fn: T, opts: CrewAIToolOptions = {}): T {
    const options: ProtectOptions = {
      toolName: opts.name ?? fn.name,
      cost: opts.cost,
      model: opts.model,
    };
    return this._guard.protect(fn, options);
  }

  /**
   * Wrap an array of CrewAI tool objects (objects with a `_run` method).
   *
   * @param tools Array of CrewAI BaseTool instances.
   * @returns     The same array with each tool's `_run` patched.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protectTools(tools: any[]): any[] {
    return tools.map((t) => this._wrapTool(t));
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private _wrapTool(tool: any): any {
    if (tool == null || typeof tool._run !== "function") return tool;

    const toolName: string = tool.name ?? tool.constructor?.name ?? "unknown_tool";
    const originalRun = tool._run.bind(tool) as AnyFn;

    tool._run = this._guard.protect(originalRun, { toolName });
    return tool;
  }
}

/**
 * Protect an entire CrewAI Crew by wrapping all agent tools.
 *
 * @param crew            A CrewAI Crew instance.
 * @param guardOrPolicy   {@link AgentGuard} or {@link AgentPolicy} to use.
 * @returns               The same Crew with all agent tools wrapped.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function protectCrew(crew: any, guardOrPolicy: AgentGuard | AgentPolicy): any {
  const crewaiGuard = new CrewAIGuard(guardOrPolicy);

  const agents: unknown[] = crew?.agents ?? [];
  for (const agent of agents) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const a = agent as any;
    if (Array.isArray(a.tools) && a.tools.length > 0) {
      a.tools = crewaiGuard.protectTools(a.tools);
    }
  }

  return crew;
}

/**
 * Protect a single CrewAI Agent by wrapping its tools.
 *
 * @param agent           A CrewAI Agent instance.
 * @param guardOrPolicy   {@link AgentGuard} or {@link AgentPolicy} to use.
 * @returns               The same Agent with its tools wrapped.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function protectCrewAIAgent(agent: any, guardOrPolicy: AgentGuard | AgentPolicy): any {
  const crewaiGuard = new CrewAIGuard(guardOrPolicy);

  if (Array.isArray(agent?.tools) && agent.tools.length > 0) {
    agent.tools = crewaiGuard.protectTools(agent.tools);
  }

  return agent;
}

File: python/agentsentinel/integrations/crewai.py
"""CrewAI integration for AgentSentinel.

CrewAI is a framework for orchestrating role-playing AI agents.
This integration wraps CrewAI tools and agent actions with AgentSentinel protection.

Usage::

    from crewai import Agent, Task, Crew
    from agentsentinel import AgentPolicy, AgentGuard
    from agentsentinel.integrations.crewai import protect_crew, CrewAIGuard

    policy = AgentPolicy(daily_budget=10.0, require_approval=["send_email"])
    guard = AgentGuard(policy)

    # Option 1: Protect an entire Crew
    crew = Crew(agents=[...], tasks=[...])
    protected_crew = protect_crew(crew, guard=guard)
    result = protected_crew.kickoff()

    # Option 2: Protect individual tools
    crewai_guard = CrewAIGuard(guard)

    @crewai_guard.tool
    def search_database(query: str) -> str:
        return db.search(query)

.. note::
    This module does **not** import ``crewai`` at the module level, so
    ``agentsentinel`` has no hard dependency on CrewAI.
"""

from __future__ import annotations

import functools
from typing import Any, Callable, Dict, List, Optional, TypeVar, Union

from ..guard import AgentGuard
from ..policy import AgentPolicy

F = TypeVar("F", bound=Callable[..., Any])


# ---------------------------------------------------------------------------
# Core guard class
# ---------------------------------------------------------------------------


class CrewAIGuard:
    """Wraps CrewAI tools and agents with :class:`.AgentGuard` policy enforcement.

    Parameters
    ----------
    guard:
        A pre-configured :class:`.AgentGuard` instance.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.
        Ignored when *guard* is provided.

    Example
    -------
    ::

        from agentsentinel import AgentGuard, AgentPolicy
        from agentsentinel.integrations.crewai import CrewAIGuard

        policy = AgentPolicy(daily_budget=5.0, require_approval=["send_email"])
        crewai_guard = CrewAIGuard(AgentGuard(policy))

        @crewai_guard.tool(cost=0.01)
        def search_web(query: str) -> str:
            return f"Results for {query}"
    """

    def __init__(
        self,
        guard: Optional[AgentGuard] = None,
        policy: Optional[AgentPolicy] = None,
    ) -> None:
        if guard is not None:
            self._guard = guard
        elif policy is not None:
            self._guard = AgentGuard(policy)
        else:
            self._guard = AgentGuard(AgentPolicy())

    @property
    def guard(self) -> AgentGuard:
        """Access the underlying :class:`.AgentGuard`."""
        return self._guard

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def tool(
        self,
        func: Optional[F] = None,
        *,
        name: Optional[str] = None,
        cost: Optional[float] = None,
        model: Optional[str] = None,
    ) -> Union[F, Callable[[F], F]]:
        """Decorator to protect a CrewAI tool function.

        Can be used with or without arguments::

            @crewai_guard.tool
            def search_web(query: str) -> str:
                return search(query)

            @crewai_guard.tool(cost=0.05, model="gpt-4o")
            def analyze_data(data: str) -> str:
                return analyze(data)

        Parameters
        ----------
        func:
            The tool function to protect (when used without parentheses).
        name:
            Override tool name (defaults to function name).
        cost:
            Fixed cost for this tool (USD).
        model:
            Model name for cost tracking.

        Returns
        -------
        Protected tool function.
        """
        def decorator(f: F) -> F:
            tool_name = name or f.__name__
            return self._guard.protect(f, tool_name=tool_name, cost=cost, model=model)  # type: ignore[return-value]

        if func is not None:
            return decorator(func)
        return decorator

    def protect_tools(self, tools: List[Any]) -> List[Any]:
        """Protect a list of CrewAI tool objects.

        Parameters
        ----------
        tools:
            List of CrewAI ``BaseTool`` instances.

        Returns
        -------
        The same list with each tool's ``_run`` method wrapped.
        """
        return [self._wrap_tool(tool) for tool in tools]

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _wrap_tool(self, tool: Any) -> Any:
        """Wrap a single CrewAI tool's ``_run`` method."""
        try:
            from crewai.tools import BaseTool  # type: ignore[import]
        except ImportError:
            # CrewAI not installed — return as-is.
            return tool

        if not isinstance(tool, BaseTool):
            return tool

        original_run = tool._run
        tool_name = getattr(tool, "name", tool.__class__.__name__)

        @functools.wraps(original_run)
        def protected_run(*args: Any, **kwargs: Any) -> Any:
            protected = self._guard.protect(
                lambda *a, **kw: original_run(*a, **kw),
                tool_name=tool_name,
            )
            return protected(*args, **kwargs)

        tool._run = protected_run
        return tool


# ---------------------------------------------------------------------------
# Convenience one-liners
# ---------------------------------------------------------------------------


def protect_crew(
    crew: Any,
    *,
    guard: Optional[AgentGuard] = None,
    policy: Optional[AgentPolicy] = None,
) -> Any:
    """Protect an entire CrewAI ``Crew`` by wrapping all agent tools.

    Parameters
    ----------
    crew:
        A ``crewai.Crew`` instance.
    guard:
        Pre-configured :class:`.AgentGuard`.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.

    Returns
    -------
    The same ``Crew`` instance with every agent's tools wrapped.

    Example
    -------
    ::

        from agentsentinel import AgentPolicy
        from agentsentinel.integrations.crewai import protect_crew

        crew = Crew(agents=[researcher, writer], tasks=[research_task, write_task])
        protected_crew = protect_crew(crew, policy=AgentPolicy(daily_budget=5.0))
        result = protected_crew.kickoff()
    """
    crewai_guard = CrewAIGuard(guard=guard, policy=policy)

    for agent in getattr(crew, "agents", []):
        if hasattr(agent, "tools") and agent.tools:
            agent.tools = crewai_guard.protect_tools(agent.tools)

    return crew


def protect_agent(
    agent: Any,
    *,
    guard: Optional[AgentGuard] = None,
    policy: Optional[AgentPolicy] = None,
) -> Any:
    """Protect a single CrewAI ``Agent`` by wrapping its tools.

    Parameters
    ----------
    agent:
        A ``crewai.Agent`` instance.
    guard:
        Pre-configured :class:`.AgentGuard`.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.

    Returns
    -------
    The same ``Agent`` with its tools wrapped.
    """
    crewai_guard = CrewAIGuard(guard=guard, policy=policy)

    if hasattr(agent, "tools") and agent.tools:
        agent.tools = crewai_guard.protect_tools(agent.tools)

    return agent
